File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque

# 사방탐색을 위해서 상우하좌 순서
dx = [0, 1, 0, -1]
dy = [-1, 0, 1, 0]

def bfs(y, x):
    global visited, miro, result
    q = deque() # 큐 생성
    q.append((y,x)) # 큐에 해당 좌표 넣고
    result[y][x] = 1 # 결과 배열에 1추가.
    while q: # 큐가 있으면
      y, x = q.popleft() # y, x 좌표 꺼내서
      for i in range(4): # 4방탐색 돌림
        idx = x + dx[i]
        idy = y + dy[i]  
        if idx >= 0 and idx < W and idy >= 0 and idy < H: # 범위 안에 있으면서
            if not visited[idy][idx] and miro[idy][idx] == 1: # 아직 방문 안 했고, 미로가 1이면
              visited[idy][idx] = True # 방문표시 하고
              q.append((idy,idx)) # 큐에 넣고
              result[idy][idx] = result[y][x] + 1 # 결과배열에 이전 위치+1로 넣어줌(거리)
                

H, W = map(int, sys.stdin.readline().strip().split())

miro = [] # 입력받을 미로
result = [] # 거리를 출력할 배열
result = [[0] * W for _ in range(H)]
visited = [[False] * W for _ in range(H)]
# 미로 입력
for i in range(H):
    miro.append(list(map(int, sys.stdin.readline().strip())))


bfs(0,0)
print(result[H-1][W-1])

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {

	static int N, M;
	static char[][] miro;
	static boolean[][] visited;
	// 상하좌우
	static int[] dx = {0, 0, -1, 1};
	static int[] dy = {-1, 1, 0, 0};

	public static void main(String[] args) throws IOException {
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

		String[] sa = bf.readLine().split(" ");
		N = Integer.parseInt(sa[0]);
		M = Integer.parseInt(sa[1]);

		visited = new boolean[N][M];
		miro = new char[N][M];
		for (int i = 0; i < N; i++) {
			char[] array = bf.readLine().toCharArray();
			for (int j = 0; j < M; j++) {
				miro[i][j] = array[j];
			}
		}

		int step = bfs();

		System.out.println(step);

	}

	public static int bfs() {
		Queue<int[]> q = new LinkedList<>();
		q.add(new int[] {0, 0, 1});
		visited[0][0] = true;

		while (!q.isEmpty()) {
			int[] temp = q.poll();
			int y = temp[0];
			int x = temp[1];
			int step = temp[2];

			if (y == N - 1 && x == M - 1) {
				return step;
			}

			for (int i = 0; i < 4; i++) {
				int ndy = y + dy[i];
				int ndx = x + dx[i];

				// 미로의 배열을 벗어나는지 확인
				if (ndy < 0 || ndx < 0 || ndy >= N || ndx >= M)
					continue;
				// 이미 도착했는지
				if (visited[ndy][ndx])
					continue;
				// 미로의 좌표가 0인지
				if (miro[ndy][ndx] == '0')
					continue;

				visited[ndy][ndx] = true;
				q.add(new int[] {ndy, ndx, step + 1});
			}
		}

		return 0;
	}
}